gordon_janitor_gcp/gce_authority.py

```diff
@@ -0,0 +1,181 @@
+            self.session,
+            tag_blacklist=tag_blacklist,
+            metadata_blacklist=metadata_blacklist)
+
+    def build_authority(self):
+        """Coordinate dependencies and create a GCEAuthority object.
+
+        Returns:
+            obj: A configured GCEAuthority object.
+        """
+        crm_client = self._get_crm_client()
+        gce_client = self._get_gce_client()
+        msg_gen = self._get_msg_generator()
+
+        return GCEAuthority(self.config, crm_client, gce_client, msg_gen,
+                            self.rrset_channel, **self.kwargs)
+
+
+def get_authority(config, rrset_channel, **kwargs):
+    builder = GCEAuthorityBuilder(config, rrset_channel, **kwargs)
+    return builder.build_authority()
+
+
+@zope.interface.implementer(interfaces.IAuthority)
+class GCEAuthority:
+    """Gordon-Janitor plugin that gathers "state of truth" GCE instance data.
+
+    Args:
+        config (dict): plugin-specific configuration.
+        crm_client (obj): client used to fetch GCE projects.
+        gce_client (obj): client used to fetch instances for each project.
+        msg_generator (obj): obj used to transform instance data into resource
+            record set messages for the reconciler.
+        rrset_channel (asyncio.Queue): channel to send resource record messages
+            to.
+    """
+    def __init__(self,
+                 config,
+                 crm_client,
+                 gce_client,
+                 msg_generator,
+                 rrset_channel=None,
+                 **kwargs):
+        self.config = config
+        self.crm_client = crm_client
+        self.gce_client = gce_client
+        self.msg_generator = msg_generator
+        self.rrset_channel = rrset_channel
+
+    async def _get_active_project_ids(self):
+        active_projects = await self.crm_client.list_all_active_projects()
+        return set(p.get('projectId') for p in active_projects)
+
+    async def _get_projects(self):
+        projects = await self._get_active_project_ids()
+        project_blacklist = set(self.config.get('project_blacklist', []))
+        # TODO: emit a metric for all projects and a metric for
+        # project - blacklist.
+        return sorted(projects - project_blacklist)
+
+    async def _get_instances(self, projects):
+        for project in projects:
+            yield await self.gce_client.list_instances(project)
+
+    async def start(self):
+        """Gather instance data batches and send them to the
+        :py:obj:`self.rrset_channel`.
+        """
+        projects = await self._get_projects()
+
+        async for instances in self._get_instances(projects):
+            rrsets = await self.msg_generator.generate_batch(instances)
+            await self.rrset_channel.put(rrsets)
+        # TODO: emit a metric of domain records created per zone.
+
+        await self.done()
+
+    async def done(self):
+        """Clean up after a run."""
+        await self.rrset_channel.put(None)
```

Comment:
You'll need to also close the session(s) for the clients that this authority uses.

Reply:
Also, here would be where the plugin would need to cleanup its tasks it created if we choose to go that route.

Reply:
Ah and a log line saying the authority is complete or something.